### PR TITLE
OSD4804: Add ServiceQuotas command routing with aliases

### PR DESCRIPTION
**Pull request by drewandersonnz**
ServiceQuotas: Add ServiceQuotas command routing with aliases
ServiceQuotas: Query for specific QuotaCode

Account CLI: Have different output formats including default, json, env

Factory: use target credentials from assumed role

### CODE REVIEW

File: cmd/account/cli.go [hunk 1]
```diff
@@ -71,20 +71,29 @@ func (o *cliOptions) complete(cmd *cobra.Command) error {
 }
 
 func (o *cliOptions) run() error {
-	awsClient, err := o.k8sclusterresourcefactory.GetCloudProvider(o.verbose)
+	_, err := o.k8sclusterresourcefactory.GetCloudProvider(o.verbose)
```
Comment (fahlmant) [inline cmd/account/cli.go:74]: Is this call still needed if we aren't using its output?

File: cmd/account/cli.go [hunk 2]
```diff
@@ -71,20 +71,29 @@ func (o *cliOptions) complete(cmd *cobra.Command) error {
...
 	}
 
-	credentials, err := awsprovider.GetAssumeRoleCredentials(awsClient, &o.k8sclusterresourcefactory.Awscloudfactory.ConsoleDuration,
-		o.k8sclusterresourcefactory.Awscloudfactory.CallerIdentity.UserId,
-		aws.String(fmt.Sprintf("arn:aws:iam::%s:role/%s",
-			o.k8sclusterresourcefactory.AccountID,
-			o.k8sclusterresourcefactory.Awscloudfactory.RoleName)))
-	if err != nil {
-		return err
+	creds := o.k8sclusterresourcefactory.Awscloudfactory.Credentials
```
Comment (fahlmant) [inline cmd/account/cli.go:79]: ah okay I see now, the GetCloudProvider assumes the right role and places the credentials here. I think that GetCloudProvider has too much logic in it then if there are cases where we don't use the awsClient it provides. There can also be cases where we want the root level creds and not assume a role.

File: cmd/account/cli.go [hunk 3]
```diff
@@ -71,20 +71,29 @@ func (o *cliOptions) complete(cmd *cobra.Command) error {
...
+
+	if o.output == "default" {
+		fmt.Fprintf(o.IOStreams.Out, "Temporary AWS Credentials:\n%s\n", creds)
+	}
+
+	if o.output == "json" {
```
Comment (fahlmant) [inline cmd/account/cli.go:85]: json and default have the same output minus a few words of preamble. Can we consolidate these?

File: cmd/account/servicequotas/cmd.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+// NewCmdServiceQuotas implements commands related to AWS service-quotas
+func NewCmdServiceQuotas(streams genericclioptions.IOStreams, flags *genericclioptions.ConfigFlags) *cobra.Command {
+	baseCmd := &cobra.Command{
+		Use:               "servicequotas",
+		Short:             "Interact with AWS service-quotas",
+		Args:              cobra.NoArgs,
+		DisableAutoGenTag: true,
+		Run:               help,
+		Aliases:           []string{"service-quotas", "service-quota"},
```
Comment (fahlmant) [inline cmd/account/servicequotas/cmd.go:16]: `servicequotas` might be good here too.

File: cmd/account/servicequotas/describe.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+func (o *describeOptions) run() error {
+	awsClient, err := o.k8sclusterresourcefactory.GetCloudProvider(o.verbose)
+	if err != nil {
+		return err
+	}
+
+	var serviceCode string = "ec2"
+	var searchMaxResults int64 = 100
```
Comment (fahlmant) [inline cmd/account/servicequotas/describe.go:85]: Why have a specific max? Shouldn't this be paginated instead?

**Review by fahlmant — COMMENTED** [review]
There's a lot of changes to the cloud factory and to the cli command that aren't related to the changes described in the title of the PR, nor the commit message. At a minimum I think the changes should be broken up by commit, but ideally in a separate PR.

File: cmd/account/cli.go (hunk 3 above)
Comment (drewandersonnz) [reply cmd/account/cli.go:85]: json == default without the pre-text. The pre-text was there before I started. Keeping that as `default` to maintain compatibility.

Personally, I wanted the `env` output for testing, the `json` is just an extra for almost zero cost.

I vote to leave it as is.

File: cmd/account/servicequotas/cmd.go (hunk 4 above)
Comment (drewandersonnz) [reply cmd/account/servicequotas/cmd.go:16]: `servicequotas` is the root command name. Should the alias also have the original name in the list?

File: cmd/account/servicequotas/describe.go (hunk 5 above)
Comment (drewandersonnz) [reply cmd/account/servicequotas/describe.go:85]: this is related to the underlying AWS calls, I would need to understand how the pagination works to continue. At the moment it seems to default to 20 results per page with 64 total results, this is just being lazy.

File: cmd/account/cli.go (hunk 1 above)
Comment (drewandersonnz) [reply cmd/account/cli.go:74]: Yes. To do the k8s calls to get the target account information and return the creds required.

File: cmd/account/cli.go (hunk 2 above)
Comment (drewandersonnz) [reply cmd/account/cli.go:79]: A fair point, is that necessary for this commit/change? Could that be a future enhancement?

**Comment by drewandersonnz** [comment]
@fahlmant the changes to the factory are to support the ServiceQuotas and improve the factory. I had a misunderstanding of how the codebase worked when branching out to create the factory and it didn't behave as expected. This solves that, while supporting the ServiceQuotas call, and backporting to `account cli` to allow for faster manual testing.

**Comment by fahlmant** [comment]
Okay. Please separate the cloudfactory changes into a separate commit at the least. CLI changes should be in their own commit as well

File: cmd/account/servicequotas/cmd.go (hunk 4 above)
Comment (fahlmant) [reply cmd/account/servicequotas/cmd.go:16]: :facepalm: nah ignore me.

File: cmd/account/servicequotas/describe.go (hunk 5 above)
Comment (fahlmant) [reply cmd/account/servicequotas/describe.go:85]: The return from ListServiceQuotas give a NextToken
https://docs.aws.amazon.com/sdk-for-go/api/service/servicequotas/#ListServiceQuotasOutput

which can be used in the ListServieQuotasInput
https://docs.aws.amazon.com/sdk-for-go/api/service/servicequotas/#ListAWSDefaultServiceQuotasInput

This patter in used in the reuse code of the accountclaim controller in the aws-account-operator, and all over the place in the aws-account-shredder

File: cmd/account/cli.go (hunk 3 above)
Comment (fahlmant) [reply cmd/account/cli.go:85]: Keeping default to maintain compatibility with what?

File: cmd/account/servicequotas/describe.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	return nil
+}
+
+func (o *describeOptions) run() error {
+	awsClient, err := o.k8sclusterresourcefactory.GetCloudProvider(o.verbose)
+	if err != nil {
+		return err
+	}
+
+	var serviceCode string = "ec2"
```
Comment (rogbas) [inline cmd/account/servicequotas/describe.go:84]: As a good practice I would define the serviceCode as a global variable in this file.

File: cmd/account/servicequotas/describe.go (hunk 5 above)
Comment (rogbas) [reply cmd/account/servicequotas/describe.go:85]: @drewandersonnz I would be ok with addressing the pagination on the 2nd PR of that card. For this initial PR this max results covers all the available ec2 quotas. 

I'm happy to have a quick sync to go over it this week.

File: cmd/account/servicequotas/describe.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+	})
+	if err != nil {
+		return err
+	}
+
+	if o.queryQuotaCode == "" {
+		fmt.Println(servicequotas.Quotas)
+	} else {
+		for _, quota := range servicequotas.Quotas {
+			if *quota.QuotaCode == o.queryQuotaCode {
```
Comment (rogbas) [inline cmd/account/servicequotas/describe.go:99]: We should be more verbose here when the specified quota is not found, outputting a msg for example.

File: cmd/account/servicequotas/describe.go (hunk 5 above)
Comment (drewandersonnz) [reply cmd/account/servicequotas/describe.go]: pagination done :-) thanks @fahlmant for the pointer to examples

File: cmd/account/servicequotas/describe.go (hunk 6 above)
Comment (drewandersonnz) [reply cmd/account/servicequotas/describe.go]: moved to CLI param instead, with default = ec2

File: cmd/account/servicequotas/describe.go (hunk 7 above)
Comment (drewandersonnz) [reply cmd/account/servicequotas/describe.go:116]: added error return when not found

**Review by iamkirkbater — COMMENTED** [review]
LGTM so far :)

File: cmd/account/cli.go (hunk 3 above)
Comment (iamkirkbater) [reply cmd/account/cli.go:85]: OSD-5676 created to address the global output and cleaning it up across the entire application.

**Pull request merged**